common/blobstore/metricClient.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package blobstore
+
+import (
+	"context"
+	"github.com/uber/cadence/common/blobstore/blob"
+	"github.com/uber/cadence/common/metrics"
+)
+
+var _ Client = (*metricClient)(nil)
+
+type metricClient struct {
+	client        Client
+	metricsClient metrics.Client
+}
+
+// NewMetricClient creates a new instance of Client that emits metrics
+func NewMetricClient(client Client, metricsClient metrics.Client) Client {
+	return &metricClient{
+		client:        client,
+		metricsClient: metricsClient,
+	}
+}
+
+func (c *metricClient) Upload(ctx context.Context, bucket string, key blob.Key, blob *blob.Blob) error {
+	c.metricsClient.IncCounter(metrics.BlobstoreClientUploadScope, metrics.CadenceClientRequests)
+
+	sw := c.metricsClient.StartTimer(metrics.BlobstoreClientUploadScope, metrics.CadenceClientLatency)
+	err := c.client.Upload(ctx, bucket, key, blob)
+	sw.Stop()
+
+	if err != nil {
+		c.metricsClient.IncCounter(metrics.BlobstoreClientUploadScope, metrics.CadenceClientFailures)
```

Comment:
Can you also emit a metric per type of error so we can also get a breakdown of different kind of errors happening for various BlobStoreClient api calls.

Reply:
I am unsure about this one for a few reasons let me know what you think 

- We do not do this in our other clients: frontend, admin, history. I like following the same pattern in blobstore.
- Some of the error types will be private repo types (e.g. lets say we cannot turn a key into a terrablob path because it breaks some constraint of terrablob). Those error types cannot be included here.
- In the happy path very few errors should be occurring if we start to see lots of errors here we will go to the logs and dive deeper into the specific errors we are getting.

Reply:
We are actually planning to emit error type metric for other clients too.  Absence of this metric caused a servicing headache recently as it could have made things simpler to debug.  
Having said that, I'm also ok if you decide to do this later once we convert all metric clients to emit error type metric.

Reply:
Okay for now I will leave it as is. But let's chat about how to do this in the future once some of the errors are private error types surfaced from Terrablob.